synse/websocket.go

```diff
@@ -0,0 +1,515 @@
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// WriteAsync writes data to a device, in an asynchronous manner.
+// NOTE - this method is not applicable for websocket client.
+func (c *websocketClient) WriteAsync(id string, opts []scheme.WriteData) (*[]scheme.Write, error) {
+	return nil, nil
+}
+
+// WriteSync writes data to a device, waiting for the write to complete.
+// FIXME - unlike http client's one, this does not support bulk write. Hence, there
+// should only be one WriteData value in the request opts. If multiple values are
+// specified, the first one would be picked. Same for the returned Transaction,
+// there should be only one value.
+// NOTE - this is related to compatibility issue mentioned in synse.go about the
+// need of uniforming client api.
+func (c *websocketClient) WriteSync(id string, opts []scheme.WriteData) (*[]scheme.Transaction, error) {
+	req := scheme.RequestWrite{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestWrite,
+		},
+		Data: opts[0], // use the first one in the slice.
+	}
+
+	resp := new(scheme.ResponseWriteState)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	// Quick fix for the issue mentioned above: return a slice with a single value.
+	out := []scheme.Transaction{}
+	out = append(out, resp.Data)
+	return &out, nil
+}
+
+// Transactions returns the sorted list of all cached transaction IDs.
+// NOTE - this method is not applicable for websocket client.
+func (c *websocketClient) Transactions() (*[]string, error) {
+	return nil, nil
+}
+
+// Transaction returns the state and status of a write transaction.
+func (c *websocketClient) Transaction(id string) (*scheme.Transaction, error) {
+	req := scheme.RequestTransaction{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestTransaction,
+		},
+		Data: scheme.WriteData{
+			Transaction: id,
+		},
+	}
+
+	resp := new(scheme.ResponseWriteState)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// GetOptions returns the current config options of the client.
+func (c *websocketClient) GetOptions() *Options {
+	return c.options
+}
+
+// addCounter safely increases the counter by 1.
+func addCounter() uint64 {
+	return atomic.AddUint64(&counter, 1)
+}
+
+// getCounter safely gets current value of counter.
+// func getCounter() uint64 {
+// 	return atomic.LoadUint64(&counter)
+// }
+
+// makeRequest issues a request event, reads its response event and parse the
+// response back in JSON.
+// TODO - how async work in this case? is writeJSON and readJSON block or
+// should it? do i need to verify returned response? how do i do that?
```

Comment:
I'll need to do some research here. I know that we will need to verify the response, against the request and it may come back out of order, so we may need to do some additional req/resp tracking at the client level.